packages/components/src/box-model-overlay/README.md

```diff
@@ -0,0 +1,158 @@
+
+```jsx
+const Example = () => {
+	const targetRef = useRef();
+
+	return (
+		<>
+			<YourComponent ref={ targetRef } />
+
+			<BoxModelOverlay
+				showValues={ showValues }
+				targetRef={ targetRef }
+			/>
+		</>
+	);
+};
+```
+
+`<BoxModelOverlay>` internally uses [`Popover`](https://github.com/WordPress/gutenberg/blob/HEAD/packages/components/src/popover/README.md) to position the overlay. This means that you can use `<Popover.Slot>` to alternatively control where the overlay is rendered.
+
+```jsx
+const Example = () => {
+	return (
+		<>
+			<BoxModelOverlay showValues={ showValues }>
+				<YourComponent />
+			</BoxModelOverlay>
+
+			<Popover.Slot />
+		</>
+	);
+};
+```
+
+`<BoxModelOverlay>` under the hood listens to size and style changes of the target element to update the overlay style automatically using `ResizeObserver` and `MutationObserver`. In some edge cases when the observers aren't picking up the changes, you can use the instance method `update` on the ref of the overlay to update it manually.
+
+```jsx
+const Example = () => {
+	const overlayRef = useRef();
+
+	// Update the overlay style manually when `deps` changes.
+	useEffect( () => {
+		overlayRef.current.update();
+	}, [ deps ] );
+
+	return (
+		<BoxModelOverlay showValues={ showValues } ref={ overlayRef }>
+			<YourComponent />
+		</BoxModelOverlay>
+	);
+};
+```
+
+Here's an example of using it with `<BoxControl>`:
+
+```jsx
+const Example = () => {
+	const [ values, setValues ] = useState( {
+		top: '50px',
+		right: '10%',
+		bottom: '50px',
+		left: '10%',
+	} );
+	const [ showValues, setShowValues ] = useState( {} );
+
+	return (
+		<>
+			<BoxControl
+				label="Padding"
+				values={ values }
+				onChange={ ( nextValues ) => setValues( nextValues ) }
+				onChangeShowVisualizer={ setShowValues }
+			/>
+
+			<BoxModelOverlay showValues={ showValues }>
+				<div
+					style={ {
+						height: 300,
+						width: 300,
+						paddingTop: values.top,
+						paddingRight: values.right,
+						paddingBottom: values.bottom,
+						paddingLeft: values.left,
+					} }
+				/>
+			</BoxModelOverlay>
+		</>
+	);
+};
+```
+
+## Props
+
+Additional props not listed below will be passed to the underlying `Popover` component.
+
+### `showValues`
+
+Controls which overlays and sides are visible. Currently the only properties supported are `margin` and `padding`, each with four sides (`top`, `right`, `bottom`, `left`).
+
+- Type: `Object`
+- Required: Yes
+- Default: `{}`
+
+### `children`
+
+A single React element to rendered as the target. It should implicitly accept `ref` to be passed in.
+
+- Type: `React.ReactElement`
+- Required: Yes if `targetRef` is not passed
+
+### `targetRef`
+
+A ref object for the target element.
+
+- Type: `Ref<HTMLElement>`
+- Required: Yes if `children` is not passed
```

Comment:
Ideally, we'd like to avoid having conditional logic in the types (see [this recent comment](https://github.com/WordPress/gutenberg/pull/40164#pullrequestreview-936468246) by @mirka ) — can we find an alternative here?

Reply:
This feels backward to me, though. I don't think we should change how a component works because the tooling doesn't support it; we should update the tooling to support it instead. Is there any other reason why we shouldn't use conditional props?

Reply:
Right, we're definitely not saying that docgen limitations should be the driving factor for designing component APIs. The general sentiment of wanting to avoid conditional props was triggered by situations where it just added to the maintenance burden. It's a slippery slope that can easily complicate itself over time into logic that is hard to hold in your head. This is nonetheless a cognitive burden that slows down dev work, no matter how well TypeScript can programatically enforce the conditions.

So it's actually not just about docgen _tooling_, but humans and human-facing docs in general, because there's still a limit to how coherently you can describe conditional props even in a handwritten README.md. It may seem simple enough now, but we also have to consider how those conditions may be forced to evolve/complicate over time, and how it adds to the cognitive load of both maintainers and consumers.

If it's absolutely worth those costs, sure! But it's good to consider whether there is a simpler, human-friendly API we can land on (e.g. remove the `children` usage, or export them as separate components).